feature-libs/product/multi-dimensional/core/services/product-multi-dimensional.service.ts

```diff
@@ -0,0 +1,122 @@
+    const levels = product.categories;
+    const code = product.code ?? '';
+
+    const variantCategories: VariantCategory[] = [];
+
+    /**
+     * Example
+     * Given at the beginning we have selected "Blue 2 L",
+     *
+     * Color:        Red             (Blue)
+     * Size:      1     2         1      (2)
+     * Fit:     M  L  M   L     M   L   M   (L)
+     *
+     * At the first level "Color", all available colors such as Red and Blue are extracted, but also
+     * elements from the Blue node are used to extract another set of options.
+     * This recursive process continues until all Variant Options have been exhaustively retrieved
+     *
+     */
+
+    levels.forEach((_, i: number) => {
+      const parentVariantCategory = variantMatrix[0].parentVariantCategory;
+
+      const variantCategory: VariantCategory = {
+        name: parentVariantCategory.name,
+        variantOptions: [],
+        hasImages: parentVariantCategory.hasImage,
+        order: parentVariantCategory.priority ?? i,
+      };
+
+      variantMatrix.forEach((element: VariantMatrixElement, j: number) => {
+        const images = this.imagesService.getVariantOptionImages(
+          element.variantOption.variantOptionQualifiers,
+          element.variantValueCategory.name
+        );
+
+        const variantOptionCategory = {
+          images,
+          value: element.variantValueCategory.name,
+          code: element.variantOption.code,
+          order: element.variantValueCategory.sequence ?? j,
+        };
+
+        variantCategory.variantOptions.push(variantOptionCategory);
+
+        if (element.variantOption.code === code && element.elements.length) {
+          /**
+           * Update the variantMatrix to reflect the currently selected elements.
+           * For example, when navigating through the Color category,
+           * adjust the matrix to represent the Size options
+           * available within the Blue branch
+           *
+           * Color:        Red             (Blue)
+           * Size:      _1_     _2_      1      (2)
+           */
+          variantMatrix = element.elements;
+        }
+      });
+
+      variantCategories.push(variantCategory);
+    });
+
+    return this.sortAndCheckIfEveryOptionHasImages(variantCategories);
+  }
+
+  /**
+   * Sorts the variant options and checks if every variant option in the category
+   * has images.
+   *
+   * Example:
+   *
+   * {
+   *   name: "Color",
+   *   variantOptions: [
+   *     {
+   *       value: Blue
+   *       images: [...]
+   *     },
+   *     {
+   *       value: Red
+   *       images: [...]
+   *     }
+   *   ]
+   * }
+   */
+  protected sortAndCheckIfEveryOptionHasImages(
+    variantCategories: VariantCategory[]
+  ): VariantCategory[] {
+    return variantCategories
+      .map((variantCategory: VariantCategory) => {
+        const variantOptions = variantCategory.variantOptions;
+        const hasImages =
+          variantCategory.hasImages ??
```

Comment:
[QUESTION]
Should we really derive the `hasImages` from the presence of images?
As we discussed in the call, the original `hasImages` property comes from the backend CMS and informs whether "should show images", but not "whether there are any images". I guess we want to treat CMS response with the value `hasImages` as a source of truth, right?

Reply:
I believe it's in the customer's best interest to present the best possible UI. If the "hasImages" flag is set to true but not every element contains an image, it would be preferable not to display any images. Utilizing the logger, we can notify the customer about which elements do not have an image. This would allow them to fix it more easily.